libbpf-tools/klockstat.c

```diff
@@ -368,25 +390,30 @@ static void print_acq_stat(struct ksyms *ksyms, struct stack_stat *ss,
 	int i;
 
 	printf("%37s %9s %8llu %10s %12s\n",
-	       symname(ksyms, ss->bt[0], buf, sizeof(buf)),
+	       print_caller(buf, sizeof(buf), ksyms, ss),
 	       print_time(avg, sizeof(avg), ss->ls.acq_total_time / ss->ls.acq_count),
 	       ss->ls.acq_count,
 	       print_time(max, sizeof(max), ss->ls.acq_max_time),
 	       print_time(tot, sizeof(tot), ss->ls.acq_total_time));
 	for (i = 1; i < nr_stack_entries; i++) {
-		if (!ss->bt[i])
+		if (!ss->bt[i] || env.per_thread)
```

Comment:
if you enforced that nr_stack_entries == 1 in arg parsing, you could skip these special cases for env.per_thread.

if you want to keep the env.per_thread check, i'd be tempted to just return from print_{acq,hld}_stat() before the stack entry parsing.

(here and below).

Reply:
or maybe have completely separate print_* functions for per-thread tracking vs per-lock (stackframe) tracking.

Reply:
Yeah, it'd be better to split.